### PR TITLE
Enhance DistinctCountThetaSketchAggregationFunction

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
## Description
Enhance `DistinctCountThetaSketchAggregationFunction`, and add the following supports:
- Support aggregating on raw values of all data types (both SV and MV)
- Support nested filter (E.g. `A = 1 AND (B = 2 OR C = 3)`)
- Support filter on all data types (both SV and MV)
- Support simple union without filters (E.g. `thetaSketch(col)`)
- Support `$0` as the default sketch (sketch without filter)

Also change the `DistinctCountRawThetaSketchAggregationFunction` to have the same usage as the `DistinctCountThetaSketchAggregationFunction` and returns base64 encoded sketch.
Fix the issue where theta-sketch parameters cannot be correctly handled in `BaseBrokerRequestHandler.updateColumnNames()`

## Release Notes
This aggregation function is still in beta version. This PR involves change on the format of data sent from server to broker, so it works only when both broker and server are upgraded to the new version.

### CODE REVIEW

**Comment by mayankshriv** [comment]
> What's the motivation to add a new aggregation function as opposed to enhancing the existing one? Is there a backward compatibility issue? If not, it adds more confusion on the user side to have multiple variations of the same aggregation functions.

We discussed offline and did some performance benchmark to ensure there is no regression. Based on the testing/benchmarking results, we decided to move the code under existing function instead of creating a new one.

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/DistinctCountThetaSketchAggregationFunction.java [hunk 1]
```diff
@@ -19,136 +19,144 @@
...
-    Preconditions.checkArgument(parametersExpression.getType() == ExpressionContext.Type.LITERAL,
-        "Last argument of DistinctCountThetaSketch must be literal (post-aggregation expression)");
-    _postAggregationExpression = QueryContextConverterUtils
-        .getExpression(CalciteSqlParser.compileToExpression(postAggregationExpression.getLiteral()));
+      _inputExpressions = Collections.singletonList(_mainExpression);
+      _includeDefaultSketch = true;
+      _filterEvaluators = Collections.emptyList();
+      _postAggregationExpression = ExpressionContext.forIdentifier(DEFAULT_SKETCH_IDENTIFIER);
+    } else {
+      // Union with post-aggregation
```
Comment (mayankshriv) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/DistinctCountThetaSketchAggregationFunction.java:118]: Just to ensure, we have test for this?

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/DistinctCountThetaSketchAggregationFunction.java [hunk 2]
```diff
@@ -184,346 +192,1063 @@ public GroupByResultHolder createGroupByResultHolder(int initialCapacity, int ma
...
+    boolean[] singleValues = new boolean[numExpressions];
+    DataType[] valueTypes = new DataType[numExpressions];
+    Object[] valueArrays = new Object[numExpressions];
+    extractValues(blockValSetMap, singleValues, valueTypes, valueArrays);
+    int numFilters = _filterEvaluators.size();
+
+    // Main expression is always index 0
+    if (valueTypes[0] != DataType.BYTES) {
+      List<UpdateSketch> updateSketches = getUpdateSketches(aggregationResultHolder);
+      if (singleValues[0]) {
```
Comment (mayankshriv) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/DistinctCountThetaSketchAggregationFunction.java:205]: Are these if/else blocks re-factorable?

**Review by mayankshriv — APPROVED** [review]
Mostly minor comments, except on `raw` version being feature compatible with `non-raw` version. Please address comments before merging.

File: pinot-core/src/main/java/org/apache/pinot/core/common/ObjectSerDeUtils.java [hunk 3]
```diff
@@ -776,6 +783,69 @@ public IdSet deserialize(ByteBuffer byteBuffer) {
...
 
+  public static final ObjectSerDe<List<Object>> LIST_SER_DE = new ObjectSerDe<List<Object>>() {
+
+    @Override
+    public byte[] serialize(List<Object> list) {
+      int size = list.size();
+
+      // Directly return the size (0) for empty list
+      if (size == 0) {
+        return new byte[Integer.BYTES];
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/common/ObjectSerDeUtils.java:794]: It is safer to not reuse this as we have no control on not modifying it

File: pinot-core/src/main/java/org/apache/pinot/core/common/ObjectSerDeUtils.java [hunk 4]
```diff
@@ -776,6 +783,69 @@ public IdSet deserialize(ByteBuffer byteBuffer) {
...
+      }
+
+      // No need to close these 2 streams
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/common/ObjectSerDeUtils.java:797]: Added

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/DistinctCountThetaSketchAggregationFunction.java [hunk 5]
```diff
@@ -19,136 +19,144 @@
...
-    }
-  }
-
-  private static final Pattern ARGUMENT_SUBSTITUTION = Pattern.compile("\\$(\\d+)");
-
-  private final ExpressionContext _thetaSketchColumn;
-  private final SetOperationBuilder _setOperationBuilder;
+@SuppressWarnings({"rawtypes", "unchecked"})
+public class DistinctCountThetaSketchAggregationFunction implements AggregationFunction<List<Sketch>, Long> {
+  private static final String SET_UNION = "SET_UNION";
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/DistinctCountThetaSketchAggregationFunction.java:81]: Using string has these 2 benefits over enum:
- Save the extra parsing of the enum
- Simplify the handling of invalid operations

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/DistinctCountThetaSketchAggregationFunction.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/DistinctCountThetaSketchAggregationFunction.java:118]: Yes, we have unit test and integration test for this

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/DistinctCountThetaSketchAggregationFunction.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/DistinctCountThetaSketchAggregationFunction.java:205]: I tried and don't see a better way to organize the code. We need to have slightly different logic for each primitive type

**Comment by Jackie-Jiang** [comment]
@mayankshriv Changed the ``DistinctCountRawThetaSketchAggregationFunction` to have the same usage as `DistinctCountThetaSketchAggregationFunction`

**Pull request merged**